hikari/impl/interaction_server.py

```diff
@@ -0,0 +1,558 @@
+
+    def __init__(
+        self,
+        status_code: int,
+        payload: typing.Optional[str] = None,
+        *,
+        content_type: typing.Optional[str] = None,
+        headers: typing.Optional[typing.MutableMapping[str, str]] = None,
+    ) -> None:
+        if payload or content_type:
+            if not headers:
+                headers = {}
+
+            headers[_CONTENT_TYPE_KEY] = content_type or _TEXT_CONTENT_TYPE
+
+        self._headers = headers
+        self._payload = payload.encode() if payload is not None else None
+        self._status_code = status_code
+
+    @property
+    def headers(self) -> typing.Optional[typing.Mapping[str, str]]:
+        return self._headers
+
+    @property
+    def payload(self) -> typing.Optional[bytes]:
+        return self._payload
+
+    @property
+    def status_code(self) -> int:
+        return self._status_code
+
+
+class InteractionServer(interaction_server.InteractionServer):
+    """Standard implementation of `hikari.api.interaction_server.InteractionServer`.
+
+    Parameters
+    ----------
+    entity_factory : hikari.api.entity_factory.EntityFactory
+        The entity factory instance this server should use.
+    event_factory : hikari.api.event_factory.EventFactory
+        The event factory instance this server should use.
+
+    Other Parameters
+    ----------------
+    dumps : aiohttp.typedefs.JSONEncoder
+        The JSON encoder this server should use. Defaults to `json.dumps`.
+    loads : aiohttp.typedefs.JSONDecoder
+        The JSON decoder this server should use. Defaults to `json.loads`.
+    public_key : builtins.bytes
+        The public key this server should use for verifying request payloads from
+        Discord. If left as `builtins.None` then the client will try to work this
+        out using `rest_client`.
+    rest_client : hikari.api.rest.RESTClient
+        The client this should use for making REST requests.
+    """
+
+    __slots__: typing.Sequence[str] = (
+        "_application_fetch_lock",
+        "_dumps",
+        "_entity_factory",
+        "_event_factory",
+        "_future",
+        "_hashes",
+        "_listeners",
+        "_loads",
+        "_rest_client",
+        "_runner",
+        "_server",
+        "_verify",
+    )
+
+    def __init__(
+        self,
+        *,
+        dumps: aiohttp.typedefs.JSONEncoder = data_binding.dump_json,
+        entity_factory: entity_factory_.EntityFactory,
+        event_factory: event_factory_.EventFactory,
+        loads: aiohttp.typedefs.JSONDecoder = data_binding.load_json,
+        rest_client: rest_client_.RESTClient,
+        public_key: typing.Optional[bytes] = None,
+    ) -> None:
+        self._application_fetch_lock = asyncio.Lock()
+        self._dumps = dumps
+        self._entity_factory = entity_factory
+        self._event_factory = event_factory
+        self._future: asyncio.Future[None] = asyncio.Future()
```

Comment:
Implementation detail wise an event is just a collection of Futures instead of just one future and that's how it'd be implemented in most scenarios so I don't really see the difference here myself but ok will change